### ISSUE TITLE
Add disjoint union to set.mm and iset.mm

### ISSUE THREAD

**Issue by jkingdon** [metamath/set.mm#2625]
As suggested in https://github.com/metamath/set.mm/issues/2595#issuecomment-1158537525 we should (presumably) add this.

The basic idea can be seen at [df-cda](http://us.metamath.org/mpeuni/df-cda.html) but the idea here is to give it some notation which makes it more useful for purposes other than cardinality.

We need three things:

* a disjoint union syntax. I guess this can be an operator in the sense of http://us.metamath.org/mpeuni/df-ov.html if we want to confine ourselves to sets. I see that the above comment refers to sets but is this usually defined on classes in general or just on sets? Suppose for the sake of argument that the notation is `( A +u B )`
* `inl` which has the property ``x e. A -> ( inl ` x ) e. ( A +u B )``
* `inr` which has the property ``x e. B -> ( inr ` x ) e. ( A +u B )``
* A theorem which lets us go the other way: ``x e. ( A +u B ) -> ( E. y e. A x = (inl ` y ) \/ E. y e. B x = ( inr ` y ) )``

Open questions:
1. sets versus proper classes as described above
2. Does defining `inl` and `inr` as functions work well? (I would imagine so by analogy with `1st`, `ZZ>=`, and others, but I haven't tried implementing anything yet)
3. Notation of the disjoint union syntax. Is this usually a plus sign as in PradicBrown2021 (from the iset.mm bibliography)? If so I suppose it should be plus sign with a subscript, which is where ASCII `+u` and typeset +<sub>u</sub> came from.

**Comment by digama0** [comment]
Common notations for disjoint union are plus, circled plus, and square cup. The definition works on proper classes as well, and the inl and inr functions can indeed be class functions which do not depend on A or B:

```
inl = ( x e. _V |-> <. (/) , x >. )
inr = ( x e. _V |-> <. 1o , x >. )
```
PS: I think df-cda is more useful with the discriminant being the first component rather than the second. [xpsc](http://us.metamath.org/mpeuni/xpsc.html) is an example of a theorem (collection) that actually makes use of this ordering since it can be used as a two-point function.

Another definition to think about is the *indexed* disjoint union, sometimes called a sigma-type. It is denoted in set.mm as `U_ x e. A ( { x } X. B )` like in [eliunxp](http://us.metamath.org/mpeuni/eliunxp.html). The plain indexed union can be written in these terms as `U_ x e. 2o ( { x } X. if ( x = (/) , A , B ) )`.

**Comment by benjub** [comment]
That would be a welcome addition.
Since this is the dual notion to the product, I would mimic what is done for (\times, \pr_1, \pr_2).
For the symbol, \sqcup and \coprod (or \amalg) are probably more standard.  I've also seen \uplus.

**Edit:** as for the first basic theorems to prove, there are the variable-free versions of those you wrote:
```
inl(A) \cap inr(B) = (/)
inl(A) \cup inr(B) = A \sqcup B
```
and of course the universal property. (The one for the product is probably already in set.mm ? At some point, @avekens was proving some universal properties, so he may know.)

**Comment by jkingdon** [comment]
> The definition works on proper classes as well,

Just to make sure we understand each other, you are saying we want:

```
df-sqcup $a ( A sqcup B ) = ( ( { (/) } X. A  ) u. ( { 1o } X. B ) )
```

and the usual equality theorems, instead of

```
df-sqcup $a sqcup = ( x e. _V , y e. _V |-> ( ( { (/) } X. x  ) u. ( { 1o } X. y ) ) )
```

which would let us use http://us.metamath.org/ileuni/oveq12d.html and the like, but would not work for proper classes.

I wanted to be explicit on the this point partly because of the recent discussion about "New definitions infrequent" and what that means in practice (I mean, each of the above involves a definition, but one is a new syntax and the other is the existing operator syntax).

**Comment by benjub** [comment]
Comparing with http://us.metamath.org/mpeuni/df-xp.html, I'd say your proposal is correct. Maybe write `{ (/) }` as `1o` to be more uniform with the second component (without knowing the definition of 1o, one knows that `1o =/= { 1o }`).

I may not have been clear when talking about the symbol: I think it could look like \sqcup or \amalg on the webpage, but the label should be related to its semantics rather than its look, so maybe `df-djuni` for the label and ` |_| ` for the token ?

**Comment by digama0** [comment]
@jkingdon Yes, that is my suggestion. Personally, I think we should be okay with using more "operator" style definitions if appropriate: a lot of the trouble with these in the old days was justifying that the syntax remains unambiguous, but I think we have a fairly solid understanding of how to avoid trouble here. The other issue with class operators is that they require eq and nf theorems, which regular definitions don't need (but on the other hand, regular definitions need a definitional expansion theorem while class definitions can sometimes just use the definition directly).

I have no opinion regarding the specific syntax to be used here, whether it is `|_|` or `sqcup` or `(+)` or something else.

**Comment by jkingdon** [comment]
> Maybe write `{ (/) }` as `1o` to be more uniform with the second component (without knowing the definition of 1o, one knows that `1o =/= { 1o }`).

If we write `( ( 1o X. A  ) u. ( { 1o } X. B ) )`  you need to know that `1o` is a singleton containing ordinal zero (`(/)`) whereas `( ( { (/) } X. A  ) u. ( { 1o } X. B ) )` has two singletons, containing zero and one. That is, it isn't enough for `{ (/) }` and `{ 1o }` to be unequal, they need to be disjoint (and are needlessly complex unless they are singletons).

**Comment by jkingdon** [comment]
> regular definitions need a definitional expansion theorem

Ooo good point.

As for the name, what I wrote before was supposed to be a placeholder rather than a proposal, but maybe the following?

```
df-djun $a ( A |_| B ) = ( ( { (/) } X. A  ) u. ( { 1o } X. B ) )

htmldef "|_|" as ' &#x2294; ';
  althtmldef "|_|" as ' &#x2294; ';
  latexdef "|_|" as "\sqcup";
```

It should be `djun` rather than `djuni` (by analogy to `df-un` versus `df-uni`).

**Comment by digama0** [comment]
Is `dju` too short? I like 3 letter label fragments.

**Comment by jkingdon** [comment]
> Is `dju` too short? I like 3 letter label fragments.

I was wondering about that too. Because we'll have `djueq12` or other longer names which include this fragment.

**Comment by benjub** [comment]
Both are fine by me.

**Comment by jkingdon** [comment]
Fixed by #2636 and #2642 (there are some follow up topics, but those can be future issues or pull requests, no need to keep this one open).
